### PR TITLE
Add a nearly-working SSL test

### PR DESCRIPTION
**Pull request by iffy**
I'd like to use BearSSL for some HTTPS requests. This PR adds a nearly-working test case. I can't figure out what's wrong with the code and was wondering:

1. Would you like to have a working SSL test in here (once it's fixed, of course)?
2. Do you know what's wrong?

I'll keep poking at it, but if you have any tips, I'd appreciate it.

Right now, it's failing with error code 31 BR_ERR_IO: https://bearssl.org/apidoc/bearssl__ssl_8h.html#a5e4e4985aec3272bbb241137b16e79af But I don't have any indication what the IO error is.

### CODE REVIEW

**Comment by arnetheduck** [comment]
https://github.com/status-im/nim-chronos/blob/master/chronos/streams/tlsstream.nim provides examples you might want to look at

File: bearssl/abi/bearssl_ssl.nim [hunk 1]
```diff
@@ -1240,18 +1240,18 @@ type
   SslioContext* {.importc: "br_sslio_context", header: "bearssl_ssl.h", bycopy.} = object
     engine* {.importc: "engine".}: ptr SslEngineContext
     lowRead* {.importc: "low_read".}: proc (readContext: pointer; data: ptr byte;
-                                        len: uint): cint {.importcFunc.}
+                                        len: uint): cint {.cdecl.}
```
Comment (arnetheduck) [inline bearssl/abi/bearssl_ssl.nim:1243]: why is this changed?

Comment (iffy) [reply bearssl/abi/bearssl_ssl.nim:1243]: I changed that because of this error:

```
test_ssl.nim(43, 14) Error: type mismatch: got <SslioContext, ptr SslEngineContext, proc (readContext: pointer, data: ptr byte, len: uint): cint{.cdecl, gcsafe, locks: 0.}, ptr SocketImpl, proc (writeContext: pointer, data: ptr byte, len: uint): cint{.cdecl, gcsafe, locks: 0.}, ptr SocketImpl>
but expected one of:
proc sslioInit(ctx: var SslioContext; engine: ptr SslEngineContext; lowRead: proc (
    readContext: pointer; data: ptr byte; len: uint): cint {.cdecl, gcsafe,
    noSideEffect, raises: [].}; readContext: pointer; lowWrite: proc (
    writeContext: pointer; data: ptr byte; len: uint): cint {.cdecl, gcsafe,
    noSideEffect, raises: [].}; writeContext: pointer)
  first type mismatch at position: 3
  required type for lowRead: proc (readContext: pointer, data: ptr byte, len: uint): cint{.cdecl, noSideEffect, gcsafe.}
  but expression 'lowRead' is of type: proc (readContext: pointer, data: ptr byte, len: uint): cint{.cdecl, gcsafe, locks: 0.}
  Pragma mismatch: got '{..}', but expected '{.noSideEffect.}'.
  This expression can have side effects. Annotate the proc with {.noSideEffect.} to get extended error information.

expression: sslioInit(io, addr(sc.eng), lowRead, addr(socket[]), lowWrite, addr(socket[]))
```

If I add `.noSideEffect.` as it suggests, then I get this error:

```
tests/test_ssl.nim(24, 10) Error: 'lowRead' can have side effects
> /Users/matt/lib/nim-bearssl/tests/test_ssl.nim(26, 20) Hint: 'lowRead' calls `.sideEffect` 'recv'
>> /Users/matt/.choosenim/toolchains/nim-1.6.10/lib/pure/net.nim(1388, 6) Hint: 'recv' called by 'lowRead'
> /Users/matt/.nimble/pkgs/unittest2-0.0.5/unittest2.nim(732, 3) Hint: 'lowRead' accesses global state 'checkpoints'
>> /Users/matt/.nimble/pkgs/unittest2-0.0.5/unittest2.nim(250, 3) Hint: 'checkpoints' accessed by 'lowRead'
```

In order to read/write to sockets, don't `lowRead` and `lowWrite` need to have side effects?

Comment (iffy) [reply bearssl/abi/bearssl_ssl.nim:1243]: I should add that this code uses the simplified I/O functions provided by BearSSL. I used https://www.bearssl.org/gitweb/?p=BearSSL;a=blob;f=samples/client_basic.c;h=d4f79fb752d5bfdb3dbc168b4d9469db11a91b7c;hb=f3ac21cb00646600475b02ecd2c592ee36e417da as an example.

**Comment by iffy** [comment]
> https://github.com/status-im/nim-chronos/blob/master/chronos/streams/tlsstream.nim provides examples you might want to look at

That looks great! Looks like that doesn't use the simplified I/O stuff. I'll try that.

File: bearssl/abi/bearssl_ssl.nim (hunk 1 above)
Comment (arnetheduck) [reply bearssl/abi/bearssl_ssl.nim:1243]: right - makes sense, meaning that an `importcproc` must be added which ensures no exceptions leak

**Comment by iffy** [comment]
Ended up using tlsstream, so I don't think this is needed anymore.
